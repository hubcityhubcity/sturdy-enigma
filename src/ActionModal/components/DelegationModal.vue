<template>
  <ActionModal
    id="delegation-modal"
    ref="actionModal"
    :validate="validateForm"
    :amount="isRedelegation ? 0 : amount"
    :title="isRedelegation ? 'Redelegate' : 'Delegate'"
    class="delegation-modal"
    submission-error-prefix="Delegating failed"
    :transaction-data="transactionData"
    :notify-message="notifyMessage"
    @close="clear"
    @txIncluded="onSuccess"
  >
    <TmFormGroup class="action-modal-form-group">
      <div class="form-message notice">
        <span v-if="!isRedelegation">
          It will take 21 days to unlock your tokens after a delegation and
          there is a risk that some tokens will be lost depending on the
          behaviour of the validator.
        </span>
        <span v-else>
          Voting power and rewards will change instantly upon redelegation â€”
          your tokens will still be subject to the risks associated with the
          original delegation for the duration of the undelegation period.
        </span>
      </div>
    </TmFormGroup>
    <TmFormGroup class="action-modal-form-group" field-id="to" field-label="To">
      <TmField
        id="to"
        v-model="targetValidator.operatorAddress"
        type="text"
        readonly
      />
      <TmFormMsg
        v-if="targetValidator.status === 'INACTIVE' && !isRedelegation"
        :msg="
          `You are about to delegate to an inactive validator (${targetValidator.statusDetailed})`
        "
        type="custom"
        class="tm-form-msg--desc"
      />
      <TmFormMsg
        v-if="targetValidator.status === 'INACTIVE' && isRedelegation"
        :msg="
          `You are about to redelegate to an inactive validator (${targetValidator.statusDetailed})`
        "
        type="custom"
        class="tm-form-msg--desc"
      />
    </TmFormGroup>

    <TmFormGroup
      v-if="fromOptions.length > 1"
      class="action-modal-form-group"
      field-id="from"
      field-label="From"
    >
      <TmField
        id="from"
        v-model="selectedIndex"
        :title="from"
        :options="fromOptions"
        type="select"
      />
    </TmFormGroup>
    <TmFormGroup
      :error="$v.amount.$error && $v.amount.$invalid"
      class="action-modal-form-group"
      field-id="amount"
      field-label="Amount"
    >
      <span class="input-suffix-denom">{{ denom }}</span>
      <TmFieldGroup>
        <TmField
          id="amount"
          v-model="amount"
          v-focus
          class="tm-field-addon"
          type="number"
          placeholder="Amount"
          @keyup.enter.native="enterPressed"
        />
        <TmBtn
          type="button"
          class="secondary addon-max"
          value="Set Max"
          @click.native="setMaxAmount()"
        />
      </TmFieldGroup>
      <span class="form-message">
        {{
          isRedelegation ? "Available to Redelegate" : "Available to Delegate"
        }}
        :
        {{ maxAmount }}
        {{ denom }}s
      </span>
      <TmFormMsg
        v-if="balance === 0"
        :msg="`doesn't have any ${denom}s`"
        name="Wallet"
        type="custom"
      />
      <TmFormMsg
        v-else-if="$v.amount.$error && !$v.amount.decimal"
        name="Amount"
        type="numeric"
      />
      <TmFormMsg
        v-else-if="$v.amount.$error && (!$v.amount.required || amount === 0)"
        name="Amount"
        type="required"
      />
      <TmFormMsg
        v-else-if="$v.amount.$error && !$v.amount.between"
        :max="$v.amount.$params.between.max"
        :min="$v.amount.$params.between.min"
        name="Amount"
        type="between"
      />
      <TmFormMsg
        v-else-if="isMaxAmount() && !isRedelegation"
        msg="You are about to use all your tokens for this transaction. Consider leaving a little bit left over to cover the network fees."
        type="custom"
        class="tm-form-msg--desc"
      />
    </TmFormGroup>
  </ActionModal>
</template>

<script>
import { mapState, mapGetters } from "vuex"
import { between, decimal } from "vuelidate/lib/validators"
import gql from "graphql-tag"
import { uatoms, SMALLEST } from "src/scripts/num"
import TmField from "src/components/common/TmField"
import TmFieldGroup from "src/components/common/TmFieldGroup"
import TmBtn from "src/components/common/TmBtn"
import TmFormGroup from "src/components/common/TmFormGroup"
import TmFormMsg from "src/components/common/TmFormMsg"
import ActionModal from "./ActionModal"
import transaction from "../utils/transactionTypes"
import { toMicroDenom } from "src/scripts/common"
import { formatBech32 } from "src/filters"
import { UserTransactionAdded } from "src/gql"

export default {
  name: `delegation-modal`,
  components: {
    TmField,
    TmFieldGroup,
    TmBtn,
    TmFormGroup,
    TmFormMsg,
    ActionModal
  },
  props: {
    targetValidator: {
      type: Object,
      required: true
    }
  },
  data: () => ({
    amount: 0,
    selectedIndex: 0,
    balance: {
      amount: 0,
      denom: ``
    },
    delegations: [],
    denom: ``
  }),
  computed: {
    ...mapState([`session`]),
    ...mapGetters([`network`, `address`]),
    fromOptions() {
      let options = [
        // from wallet
        {
          address: this.address,
          maximum: Number(this.balance.amount),
          key: `My Wallet - ${formatBech32(this.address, false, 20)}`,
          value: 0
        }
      ]

      options = options.concat(
        this.delegations
          // exclude the validator we are delegating to as a source
          .filter(
            delegation =>
              delegation.validator.operatorAddress !=
              this.targetValidator.operatorAddress
          )
          .map((delegation, index) => {
            return {
              address: delegation.validator.operatorAddress,
              maximum: Number(delegation.amount),
              key: `${delegation.validator.name} - ${formatBech32(
                delegation.validator.operatorAddress,
                false,
                20
              )}`,
              value: index + 1
            }
          })
      )

      return options
    },
    from() {
      if (!this.session.signedIn) return ``

      return this.fromOptions[this.selectedIndex].address
    },
    isRedelegation() {
      return this.from !== this.address
    },
    transactionData() {
      if (!this.from) return {}

      if (this.isRedelegation) {
        return {
          type: transaction.REDELEGATE,
          validatorSourceAddress: this.from,
          validatorDestinationAddress: this.targetValidator.operatorAddress,
          amount: uatoms(this.amount),
          denom: toMicroDenom(this.denom)
        }
      } else {
        return {
          type: transaction.DELEGATE,
          validatorAddress: this.targetValidator.operatorAddress,
          amount: uatoms(this.amount),
          denom: toMicroDenom(this.denom)
        }
      }
    },
    notifyMessage() {
      if (this.isRedelegation) {
        return {
          title: `Successful redelegation!`,
          body: `You have successfully redelegated your ${this.denom}s`
        }
      } else {
        return {
          title: `Successful delegation!`,
          body: `You have successfully delegated your ${this.denom}s`
        }
      }
    },
    maxAmount() {
      return this.fromOptions[this.selectedIndex].maximum
    }
  },
  mounted() {
    this.$apollo.queries.balance.refetch()
    this.$apollo.queries.delegations.refetch()
  },
  methods: {
    open(options) {
      if (options && options.redelegation && this.fromOptions.length > 1) {
        this.selectedIndex = 1
      }
      this.$refs.actionModal.open()
    },
    validateForm() {
      this.$v.$touch()

      return !this.$v.$invalid
    },
    clear() {
      this.$v.$reset()

      this.selectedIndex = 0
      this.amount = 0
    },
    setMaxAmount() {
      this.amount = this.maxAmount
    },
    isMaxAmount() {
      return parseFloat(this.amount) === parseFloat(this.maxAmount)
    },
    enterPressed() {
      this.$refs.actionModal.validateChangeStep()
    },
    onSuccess(event) {
      this.$emit(`success`, event)
    }
  },
  validations() {
    return {
      amount: {
        required: x => !!x && x !== `0`,
        decimal,
        between: between(SMALLEST, this.maxAmount)
      }
    }
  },
  apollo: {
    delegations: {
      query: gql`
        query DelegationsDelegationModal(
          $networkId: String!
          $delegatorAddress: String!
        ) {
          delegations(
            networkId: $networkId
            delegatorAddress: $delegatorAddress
          ) {
            amount
            validator {
              operatorAddress
              name
            }
          }
        }
      `,
      skip() {
        return !this.address
      },
      variables() {
        return {
          networkId: this.network,
          delegatorAddress: this.address
        }
      },
      update(data) {
        /* istanbul ignore next */
        return data.delegations
      }
    },
    balance: {
      query: gql`
        query BalanceDelegationModal(
          $networkId: String!
          $address: String!
          $denom: String!
        ) {
          balance(networkId: $networkId, address: $address, denom: $denom) {
            amount
            denom
          }
        }
      `,
      skip() {
        return !this.address
      },
      variables() {
        return {
          networkId: this.network,
          address: this.address,
          denom: this.denom
        }
      }
    },
    denom: {
      query: gql`
        query NetworksDelegationModal($networkId: String!) {
          network(id: $networkId) {
            id
            stakingDenom
          }
        }
      `,
      variables() {
        return {
          networkId: this.network
        }
      },
      update(data) {
        /* istanbul ignore next */
        return data.network.stakingDenom
      }
    }
  },
  $subscribe: {
    userTransactionAdded: {
      variables() {
        return {
          networkId: this.network,
          address: this.address
        }
      },
      skip() {
        return !this.address
      },
      query: UserTransactionAdded,
      result({ data }) {
        if (data.userTransactionAdded.success) {
          this.$apollo.queries.delegations.refetch()
        }
      }
    }
  }
}
</script>
