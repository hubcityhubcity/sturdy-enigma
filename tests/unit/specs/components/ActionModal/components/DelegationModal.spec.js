"use strict"

import { shallowMount, createLocalVue } from "@vue/test-utils"
import DelegationModal from "src/ActionModal/components/DelegationModal"
import Vuelidate from "vuelidate"

const validators = [
  {
    operatorAddress: "cosmosvaladdr12324536463",
    status: "ACTIVE"
  },
  {
    operatorAddress: "cosmosvaladdr1sdsdsd123123",
    status: "ACTIVE"
  },
  {
    operatorAddress: "cosmosvaladdr1kjisjsd862323",
    status: "INACTIVE",
    statusDetailed: "temporally banned from the network"
  },
  {
    operatorAddress: "cosmosvaladdr1sd0f8mnbjb2",
    status: "INACTIVE",
    statusDetailed: "banned from the network"
  }
]

describe(`DelegationModal`, () => {
  let wrapper
  const localVue = createLocalVue()
  localVue.use(Vuelidate)
  localVue.directive("focus", () => {})

  const state = {
    session: {
      signedIn: true
    }
  }

  const getters = {
    network: "testnet",
    address: "cosmos1234"
  }

  beforeEach(() => {
    wrapper = shallowMount(DelegationModal, {
      localVue,
      mocks: {
        $store: { getters, state },
        $apollo: {
          queries: {
            balance: { refetch: () => {} },
            delegations: { refetch: () => {} }
          }
        }
      },
      propsData: {
        targetValidator: validators[0]
      }
    })
    wrapper.setData({
      delegations: [
        {
          validator: validators[0],
          amount: 10
        },
        {
          validator: validators[1],
          amount: 124
        }
      ],
      denom: "STAKE",
      balance: {
        amount: 1000,
        denom: "STAKE"
      }
    })
  })

  it(`should display the delegation modal form`, async () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it(`should submit when enterPressed is called`, async () => {
    const self = {
      $refs: { actionModal: { validateChangeStep: jest.fn() } }
    }
    DelegationModal.methods.enterPressed.call(self)
    expect(self.$refs.actionModal.validateChangeStep).toHaveBeenCalled()
  })

  it(`opens`, () => {
    const $refs = { actionModal: { open: jest.fn() } }
    DelegationModal.methods.open.call({ $refs })
    expect($refs.actionModal.open).toHaveBeenCalled()
  })

  it(`opens and switches to redelegaion when selected`, () => {
    wrapper.vm.$refs = { actionModal: { open: jest.fn() } }
    wrapper.vm.open({ redelegation: true })
    expect(wrapper.vm.selectedIndex).toBe(1)
  })

  it(`clears on close`, () => {
    const self = {
      $v: { $reset: jest.fn() },
      selectedIndex: 1,
      amount: 10
    }
    DelegationModal.methods.clear.call(self)
    expect(self.$v.$reset).toHaveBeenCalled()
    expect(self.selectedIndex).toBe(0)
    expect(self.amount).toBe(0)
  })

  describe(`if amount field max button clicked`, () => {
    it(`amount has to be 1000 atom`, async () => {
      wrapper.vm.setMaxAmount()
      expect(wrapper.vm.amount).toBe(1000)
    })
  })

  describe(`validation`, () => {
    describe(`fails`, () => {
      it(`with default values`, () => {
        expect(wrapper.vm.validateForm()).toBe(false)
      })

      it(`if the user manually inputs a number greater than the balance`, () => {
        wrapper.setData({ amount: 1420 })
        expect(wrapper.vm.validateForm()).toBe(false)
      })
    })

    describe(`succeeds`, () => {
      it(`if the amount is positive and the user has enough balance`, () => {
        wrapper.setData({ amount: 50 })
        expect(wrapper.vm.validateForm()).toBe(true)
      })
    })
  })

  describe("Submission Data for Delegating", () => {
    beforeEach(() => {
      wrapper.setProps({
        targetValidator: validators[1]
      })
      wrapper.setData({
        amount: 10,
        selectedIndex: 0
      })
    })

    it("should return correct transaction data for delegating", () => {
      expect(wrapper.vm.transactionData).toEqual({
        type: "MsgDelegate",
        validatorAddress: `cosmosvaladdr1sdsdsd123123`,
        amount: "10000000",
        denom: "stake"
      })
    })

    it("should return correct notification message for delegating", () => {
      expect(wrapper.vm.notifyMessage).toEqual({
        title: `Successful delegation!`,
        body: `You have successfully delegated your STAKEs`
      })
    })
  })

  describe("Submission Data for Redelegating", () => {
    beforeEach(() => {
      wrapper.setProps({
        targetValidator: validators[1]
      })
      wrapper.setData({
        amount: 10,
        selectedIndex: 1
      })
    })

    it("should return correct transaction data for delegating", () => {
      expect(wrapper.vm.transactionData).toEqual({
        type: "MsgRedelegate",
        validatorSourceAddress: "cosmosvaladdr12324536463",
        validatorDestinationAddress: "cosmosvaladdr1sdsdsd123123",
        amount: "10000000",
        denom: "stake"
      })
      // expect(wrapper.vm.transactionData).toEqual()
    })

    it("should return correct notification message for delegating", () => {
      expect(wrapper.vm.notifyMessage).toEqual({
        title: `Successful redelegation!`,
        body: `You have successfully redelegated your STAKEs`
      })
    })
  })

  describe(`if amount field max button clicked`, () => {
    it(`amount has to be 1000 atom`, async () => {
      wrapper.setData({
        amount: 1,
        selectedIndex: 0
      })
      wrapper.vm.setMaxAmount()
      expect(wrapper.vm.amount).toBe(1000)
    })
    it(`should show warning message`, async () => {
      wrapper.setData({
        amount: 1000,
        selectedIndex: 0
      })
      //await wrapper.vm.$nextTick()
      expect(wrapper.html()).toContain(
        "You are about to use all your tokens for this transaction. Consider leaving a little bit left over to cover the network fees."
      )
    })
  })

  describe(`if validator is jailed`, () => {
    it(`must show warn message about it`, async () => {
      wrapper.setProps({
        targetValidator: validators[2] // Jailed validator
      })
      wrapper.setData({
        amount: 1,
        selectedIndex: 0
      })
      expect(wrapper.html()).toContain(
        "You are about to delegate to an inactive validator (temporally banned from the network)"
      )
    })
  })

  describe(`if validator is tombstoned`, () => {
    it(`must show warn message about it`, async () => {
      wrapper.setProps({
        targetValidator: validators[3] // Tombstoned validator
      })
      wrapper.setData({
        amount: 1,
        selectedIndex: 0
      })
      expect(wrapper.html()).toContain(
        "You are about to delegate to an inactive validator (banned from the network)"
      )
    })
  })

  describe(`if validator is active`, () => {
    it(`must not show warn message`, async () => {
      wrapper.setData({
        amount: 1,
        selectedIndex: 0,
        targetValidator: validators[0] // Active validator
      })
      expect(wrapper.html()).not.toContain(
        "You are about to delegate to an inactive validator"
      )
      expect(wrapper.find("#to .tm-form-msg--desc").exists()).toBe(false)
    })
  })
})
