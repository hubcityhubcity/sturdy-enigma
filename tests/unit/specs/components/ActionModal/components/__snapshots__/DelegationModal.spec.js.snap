// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DelegationModal should display the delegation modal form 1`] = `
<actionmodal-stub
  amount="0"
  class="delegation-modal"
  id="delegation-modal"
  notifymessage="[object Object]"
  rewards=""
  submissionerrorprefix="Delegating failed"
  title="Delegate"
  transactiondata="[object Object]"
  validate="function () { [native code] }"
>
  <tmformgroup-stub
    class="action-modal-form-group"
  >
    <div
      class="form-message notice"
    >
      <span>
        
        It will take 21 days to unlock your tokens after a delegation and
        there is a risk that some tokens will be lost depending on the
        behaviour of the validator.
      
      </span>
    </div>
  </tmformgroup-stub>
   
  <tmformgroup-stub
    class="action-modal-form-group"
    fieldid="to"
    fieldlabel="To"
  >
    <tmfield-stub
      id="to"
      readonly=""
      type="text"
      value="cosmosvaladdr12324536463"
    />
     
    <!---->
     
    <!---->
  </tmformgroup-stub>
   
  <tmformgroup-stub
    class="action-modal-form-group"
    fieldid="from"
    fieldlabel="From"
  >
    <tmfield-stub
      id="from"
      options="[object Object],[object Object]"
      title="cosmos1234"
      type="select"
      value="0"
    />
  </tmformgroup-stub>
   
  <tmformgroup-stub
    class="action-modal-form-group"
    fieldid="amount"
    fieldlabel="Amount"
  >
    <span
      class="input-suffix-denom"
    >
      STAKE
    </span>
     
    <tmfieldgroup-stub>
      <tmfield-stub
        class="tm-field-addon"
        id="amount"
        placeholder="Amount"
        type="number"
        value="0"
      />
       
      <tmbtn-stub
        class="secondary addon-max"
        type="button"
        value="Set Max"
      />
    </tmfieldgroup-stub>
     
    <span
      class="form-message"
    >
      
      Available to Delegate
      :
      1000
      STAKEs
    
    </span>
     
    <!---->
  </tmformgroup-stub>
</actionmodal-stub>
`;
